# Funnel Analysis

In [1]:
%matplotlib inline
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings('ignore')
plt.style.use('seaborn-paper')

## 1.Data load
- 문서 앱내 사용시 발생하는 클라이언 로그 데이터 
(after parsed)
- 서버로그와 달리, 유저의 행동이 발생할 경우 로그 수집 (유저 행동 패턴 파악에 용이)
    - 클라이언트(앱) 로그
    - 서버 로그

In [6]:
%cd C:\dev\github\analysis practice\로그 데이터 분석\

C:\dev\github\analysis practice\로그 데이터 분석


In [14]:
df = pd.read_csv('datas/tutorial_02/df_funnel.csv',index_col=0)
#index_col = 0 : unnamed:0 컬럼 제거

### 1.2 Columns Info.
- actiontype: 문서 이용시 행동(OPEN, CLOSE, SAVE,,)
- ismydoc: 내문서 해당 여부(중요x)
- ext: 문서 확장자
- sessionid: 유저 식별자
- documentposition: 문서 이용시 위치 정보(CLOUD, OTHERAPP)
- datetime: Timestamp
- screen: 앱내 화면 이름

In [15]:
df.head()

actiontype ismydoc  ext                         sessionid documentposition   
0       OPEN  NoView  PDF  9400fd2e43d7dc2d054ca78806236ee1     LOCALSTORAGE  \
1      CLOSE  NoView  PDF  9400fd2e43d7dc2d054ca78806236ee1     LOCALSTORAGE   
2       OPEN    View  PDF  9400fd2e43d7dc2d054ca78806236ee1   MYPOLARISDRIVE   
3      CLOSE    View  PDF  9400fd2e43d7dc2d054ca78806236ee1   MYPOLARISDRIVE   
4       OPEN  NoView  PDF  f191063c562691041dfa935ff0876975         OTHERAPP   

    datetime   screen  
0  2016.7.18  Per_Dir  
1  2016.7.18  Per_Dir  
2  2016.7.18  Pub_Dir  
3  2016.7.18  Pub_Dir  
4   2016.7.6     Main

In [16]:
# key info => data entires, missing values, type
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 301861 entries, 0 to 301860
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   actiontype        301861 non-null  object
 1   ismydoc           301861 non-null  object
 2   ext               301861 non-null  object
 3   sessionid         301861 non-null  object
 4   documentposition  301861 non-null  object
 5   datetime          301861 non-null  object
 6   screen            301861 non-null  object
dtypes: object(7)
memory usage: 18.4+ MB


data entries : 301861 / missing values : None / type : all object
> change datetime type

## 2. Preprocessing
- 데이터 타입 변환
- 데이터 값 변경
- 결측치 처리
- 신규 session id 부여
- 대소문자 처리

### 2.1 날짜를 pandas datetime 형태로 변환하기
[Reference](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.to_datetime.html)

```python
pandas.to_datetime()
```

case 1. data type is string

In [20]:
#size() : NaN 포함 , series type 반환
#count() : NaN 제외 , dataframe 반환
df.groupby('datetime').size().head(15)


datetime
2016.7.1     11836
2016.7.10     6750
2016.7.11    10075
2016.7.12    12340
2016.7.13    12159
2016.7.14    11721
2016.7.15    11057
2016.7.16     8513
2016.7.17     6906
2016.7.18     9987
2016.7.19    11872
2016.7.2      8884
2016.7.20    11912
2016.7.21    11599
2016.7.22    11250
dtype: int64

In [21]:
str_date = ['2018/01/01', '2018.01.02', '2018-01-03', '2018-01-04', '2018/01/05'] # string type
str_date

['2018/01/01', '2018.01.02', '2018-01-03', '2018-01-04', '2018/01/05']

In [26]:
#str -> datetime type -> pd.series
pd.Series(pd.to_datetime(str_date,format='mixed')) 

0   2018-01-01
1   2018-01-02
2   2018-01-03
3   2018-01-04
4   2018-01-05
dtype: datetime64[ns]

case 2. data type is timestamp 

In [34]:
ts_date = [1349720105, 1349806505, 1349892905, 1349979305, 1350065705]
ts_date

[1349720105, 1349806505, 1349892905, 1349979305, 1350065705]

In [38]:
pd.Series(pd.to_datetime(ts_date,unit='s')).dt.date

0    2012-10-08
1    2012-10-09
2    2012-10-10
3    2012-10-11
4    2012-10-12
dtype: object

case 3. change type only

In [41]:
df.datetime.info() #object -> str

<class 'pandas.core.series.Series'>
Index: 301861 entries, 0 to 301860
Series name: datetime
Non-Null Count   Dtype 
--------------   ----- 
301861 non-null  object
dtypes: object(1)
memory usage: 4.6+ MB


In [43]:
#option1. string to datetime 
df['datetime'].astype('datetime64[ns]').head()

0   2016-07-18
1   2016-07-18
2   2016-07-18
3   2016-07-18
4   2016-07-06
Name: datetime, dtype: datetime64[ns]

In [44]:
# option2.
pd.to_datetime(df['datetime']).head()

0   2016-07-18
1   2016-07-18
2   2016-07-18
3   2016-07-18
4   2016-07-06
Name: datetime, dtype: datetime64[ns]

In [47]:
df['datetime'] = pd.to_datetime(df['datetime'])

In [48]:
df.datetime.info()

<class 'pandas.core.series.Series'>
Index: 301861 entries, 0 to 301860
Series name: datetime
Non-Null Count   Dtype         
--------------   -----         
301861 non-null  datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 4.6 MB


In [49]:
df.head()

actiontype ismydoc  ext                         sessionid documentposition   
0       OPEN  NoView  PDF  9400fd2e43d7dc2d054ca78806236ee1     LOCALSTORAGE  \
1      CLOSE  NoView  PDF  9400fd2e43d7dc2d054ca78806236ee1     LOCALSTORAGE   
2       OPEN    View  PDF  9400fd2e43d7dc2d054ca78806236ee1   MYPOLARISDRIVE   
3      CLOSE    View  PDF  9400fd2e43d7dc2d054ca78806236ee1   MYPOLARISDRIVE   
4       OPEN  NoView  PDF  f191063c562691041dfa935ff0876975         OTHERAPP   

    datetime   screen  
0 2016-07-18  Per_Dir  
1 2016-07-18  Per_Dir  
2 2016-07-18  Pub_Dir  
3 2016-07-18  Pub_Dir  
4 2016-07-06     Main

### 2.2 Missing value 확인

In [52]:
df.isnull().sum()

actiontype          0
ismydoc             0
ext                 0
sessionid           0
documentposition    0
datetime            0
screen              0
dtype: int64

### 2.3 결측치 처리

- **Drop**
    - 경우에 따라 결측치 처리 방법이 달라진다.
    - 샘플수가 많다면 missing values 를 포함하는 행을 모두 삭제하는 것이 가능하다

```python
# 결측치가 하나라도 있으면 
df.dropna()

# 모든 값이 Null일 경우만
df.dropna(how='all')

# 결측치가 하나 이상 있는 Case만 선택
df[df.isnull().any(axis=1)]
```

- **Imputation**
    - 만약 샘플수가 충분하지 않을 경우, Pandas의 fillna() 명령어로 Null 값을 채우는 것이 가능하다. 
    - 연속형인 경우 Mean이나 Median을 이용하고 명목형인 경우 Mode(최빈치)나 예측 모형을 통해 Null 값을 대체할 수 있다.

```python
# Null 값을 median으로 대체하는 코드 예제
df.fillna(df.mean()) 
```

In [70]:
#날짜별 screen 사용자 이용 현황
df_by_screen = df.groupby(['datetime','screen'])['sessionid'].nunique().unstack()
df_by_screen[:10]

#.unstack() : 가장 안쪽에 있는 row를 column으로 피벗시킨 dataframe 반환 

screen      InProduct_Mob  InProduct_Web    Main  Per_Dir  Pub_Dir   
datetime                                                             
2016-07-01          137.0          147.0  3582.0    296.0   1008.0  \
2016-07-02           89.0          124.0  2683.0    206.0    801.0   
2016-07-03           50.0           83.0  1886.0    165.0    590.0   
2016-07-04          112.0          104.0  2808.0    281.0    764.0   
2016-07-05          118.0          138.0  3335.0    300.0    842.0   
2016-07-06          138.0          145.0  3256.0    293.0    855.0   
2016-07-07          110.0          117.0  3150.0    266.0    807.0   
2016-07-08          117.0          148.0  3208.0    273.0    865.0   
2016-07-09          102.0          138.0  2381.0    229.0    748.0   
2016-07-10           62.0          122.0  1872.0    230.0    636.0   

screen      Purchase_done  Purchase_page  
datetime                                  
2016-07-01            2.0           23.0  
2016-07-02            2.0           15.0  
2016-07-03            NaN           28.0  
2016-07-04            NaN           29.0  
2016-07-05            1.0           26.0  
2016-07-06            1.0           30.0  
2016-07-07            NaN           18.0  
2016-07-08            2.0           20.0  
2016-07-09            2.0           23.0  
2016-07-10            2.0           18.0

In [71]:
df_by_screen.isnull().sum()

screen
InProduct_Mob    0
InProduct_Web    0
Main             0
Per_Dir          0
Pub_Dir          0
Purchase_done    6
Purchase_page    0
dtype: int64

In [73]:
# 해당 결측치는 0으로 대체 
df_by_screen.fillna(0,inplace=True)

In [74]:
df_by_screen.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 30 entries, 2016-07-01 to 2016-07-30
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   InProduct_Mob  30 non-null     float64
 1   InProduct_Web  30 non-null     float64
 2   Main           30 non-null     float64
 3   Per_Dir        30 non-null     float64
 4   Pub_Dir        30 non-null     float64
 5   Purchase_done  30 non-null     float64
 6   Purchase_page  30 non-null     float64
dtypes: float64(7)
memory usage: 1.9 KB


### 2.4 결측치가 카테고리 변수인 경우
- Option 1. 최빈치(mode)
    - 쉽고 빠르게 이용가능하나, 낮은 적절성
- Option 2. 예측모형으로 예측치
    - 적절성은 비교적 높으나 별도의 모형과 계산량 필요

option1. 최빈값으로 대체

In [76]:
#결측치 생성
df_ms = df.copy()

df_ms.loc[2, 'documentposition'] = np.nan 
df_ms.loc[5, 'documentposition'] = np.nan 
df_ms.loc[7, 'documentposition'] = np.nan 

In [77]:
df_ms.head(10)

actiontype ismydoc  ext                         sessionid documentposition   
0       OPEN  NoView  PDF  9400fd2e43d7dc2d054ca78806236ee1     LOCALSTORAGE  \
1      CLOSE  NoView  PDF  9400fd2e43d7dc2d054ca78806236ee1     LOCALSTORAGE   
2       OPEN    View  PDF  9400fd2e43d7dc2d054ca78806236ee1              NaN   
3      CLOSE    View  PDF  9400fd2e43d7dc2d054ca78806236ee1   MYPOLARISDRIVE   
4       OPEN  NoView  PDF  f191063c562691041dfa935ff0876975         OTHERAPP   
5      RESET    View  PDF  f191063c562691041dfa935ff0876975              NaN   
6      CLOSE  NoView  PDF  3747719d8f6bdd334703320433928d5a         OTHERAPP   
7       OPEN  NoView  PDF  3747719d8f6bdd334703320433928d5a              NaN   
8       OPEN  NoView  PDF  3da5ab986c93803de1e25012d9972274         OTHERAPP   
9      CLOSE  NoView  PDF  3da5ab986c93803de1e25012d9972274         OTHERAPP   

    datetime   screen  
0 2016-07-18  Per_Dir  
1 2016-07-18  Per_Dir  
2 2016-07-18  Pub_Dir  
3 2016-07-18  Pub_Dir  
4 2016-07-06     Main  
5 2016-07-06     Main  
6 2016-07-20     Main  
7 2016-07-20     Main  
8 2016-07-28     Main  
9 2016-07-28     Main

# 최빈값 계산
